File: transformerViT/hello_parallel.c
#include <mpi.h>
#include <stdio.h>
#include <stdlib.h>

int main(int argc, char** argv) {
    MPI_Init(&argc, &argv);
    int rank, world_size;
    int values[4] = {39, 72, 129, 42};
    int scattered_val;
    MPI_Comm_size(MPI_COMM_WORLD, &world_size);
    MPI_Comm_rank(MPI_COMM_WORLD, &rank);
    MPI_Scatter(&values, 1, MPI_INT, &scattered_val, 1, MPI_INT, 0, MPI_COMM_WORLD);
    char *s_or_p_processor = world_size < 2 ? "processor" : "processors";
    printf("Processor %d received %d\n",rank, scattered_val);
    MPI_Barrier(MPI_COMM_WORLD);
    // a computation to simulate a forward and backward pass 
    scattered_val = scattered_val * 0.5; 
    printf("Processor %d updated value %d\n",rank, scattered_val);
    MPI_Barrier(MPI_COMM_WORLD);
    // gather to rank zero
    int global_avg;
    MPI_Reduce(&scattered_val, &global_avg, 1, MPI_INT, MPI_SUM, 0, MPI_COMM_WORLD);
    if (rank == 0){
        printf("Rank %d: sum %d average %.1f\n", 
            rank, global_avg, (float)global_avg/world_size);
    }
    // Broadcast the result to all the processes
    float broadcast_val = 1.0f;
    if (rank == 0){
        broadcast_val = (float)global_avg/world_size;
    }
    MPI_Bcast(&broadcast_val, 1, MPI_FLOAT, 0, MPI_COMM_WORLD);
    printf("Processor %d final value %f\n",rank, broadcast_val);
    MPI_Finalize();
}

File: transformerViT/hello_parallel.py
from mpi4py import MPI
import numpy as np 

comm = MPI.COMM_WORLD
size = comm.Get_size()
rank = comm.Get_rank()

if rank == 0:
    data = [39, 72, 129, 42]
else:
    data = None 

# simulate sending batches to nodes
data = comm.scatter(data, root=0)
print(f"Processor {rank} received {data}")
comm.Barrier() #here only for printing convenience
# simulate forward and back pass 
data = data * 0.5
print(f"Processor {rank} updated {data:.2f}")
# reduce 
resulting_sum = np.array(0.0,dtype=np.float64)
comm.Reduce(
        np.array(data, dtype=np.float64),
        resulting_sum, 
        op=MPI.SUM, root=0)

if rank == 0:
    print(f"Processor {rank} sum {resulting_sum:.2f}")
# broadcast the average
data = comm.bcast(resulting_sum/size, root=0)
print(f"Processor {rank} final {data:.2f}")

